runtime/runtime/src/bandwidth_scheduler/mod.rs

```diff
@@ -38,55 +48,89 @@ pub fn run_bandwidth_scheduler(
-            BandwidthSchedulerState { mock_data: [0; 32] }
+            BandwidthSchedulerState {
+                link_allowances: Vec::new(),
+                sanity_check_hash: CryptoHash::default(),
+            }
         }
     };
 
-    let scheduler_state_hash: CryptoHash = hash(&borsh::to_vec(&scheduler_state).unwrap());
+    // Prepare the status info for each shard based on congestion info.
+    let mut shards_status: BTreeMap<ShardId, ShardStatus> = BTreeMap::new();
+    for (shard_id, extended_congestion_info) in apply_state.congestion_info.iter() {
+        let last_chunk_missing = extended_congestion_info.missed_chunks_count > 0;
+        let is_allowed_sender_shard =
+            *shard_id == ShardId::from(extended_congestion_info.congestion_info.allowed_shard());
+
+        let congestion_level = CongestionControl::new(
+            apply_state.config.congestion_control_config,
+            extended_congestion_info.congestion_info,
+            extended_congestion_info.missed_chunks_count,
+        )
+        .congestion_level();
+        let is_fully_congested = CongestionControl::is_fully_congested(congestion_level);
+
+        shards_status.insert(
+            *shard_id,
+            ShardStatus { last_chunk_missing, is_allowed_sender_shard, is_fully_congested },
+        );
+    }
 
+    // Prepare lists of sender and receiver shards.
+    // TODO(bandwidth_scheduler) - find a better way to get the sender and receiver shards.
+    // Maybe pass shard layout in ApplyState? That might also be needed for resharding.
+    // Taking all shards from the congestion info will be wrong for only one height during
+    // resharding, so it might be good enough for now, but it's not ideal.
     let mut all_shards = apply_state.congestion_info.all_shards();
     all_shards.sort();
     if all_shards.is_empty() {
-        all_shards = vec![ShardId::new(0)];
+        // In some tests there's no congestion info and the list of shards ends up empty.
+        // Pretend that there's only one default shard.
+        all_shards = vec![ShardUId::single_shard().shard_id()];
+        shards_status.insert(
+            all_shards[0],
+            ShardStatus {
+                last_chunk_missing: false,
+                is_allowed_sender_shard: true,
+                is_fully_congested: false,
+            },
+        );
     }
-    let num_shards =
-        NonZeroU64::new(all_shards.len().try_into().expect("Can't convert usize to u64"))
-            .expect("Number of shards can't be zero");
+    let sender_shards = &all_shards;
+    let receiver_shards = &all_shards;
 
-    let params = BandwidthSchedulerParams::new(num_shards, &apply_state.config);
+    // Calculate the current scheduler parameters.
+    let num_shards: u64 = std::cmp::max(sender_shards.len(), receiver_shards.len())
+        .try_into()
+        .expect("Can't convert usize to u64");
```

Comment:
Why do you do max here? Those two variables are the same and equal to all_shards aren't they?

Reply:
Currently `senders` and `receivers` are equal, but in theory they could be different on a resharding boundary.
For example senders could be `[0, 1, 2]`, but the receipts will be received at the next height, where the shard layout will be different, e,g `[0, 1, 3, 4]` (if 2 splits into 3 and 4).

`senders` and `receivers` is a bit more future proof if we ever wanted to handle resharding properly in the future.
Although I guess even if `senders` and `receivers` are set correctly, the params will be slightly wrong after resharding...

Reply:
> Although I guess even if senders and receivers are set correctly, the params will be slightly wrong during resharding...

But it's possible to get the right params for decoding previous requests and encoding new ones, it can be done in the future if we decide to fully support resharding in the scheduler.

Reply:
Removed separate list of senders and receivers, the problem is gone (for now)